src/System.Private.CoreLib/shared/System/Security/SecureString.cs

```diff
@@ -178,19 +358,122 @@ internal unsafe IntPtr MarshalToString(bool globalAlloc, bool unicode)
             lock (_methodLock)
             {
                 EnsureNotDisposed();
-                return MarshalToStringCore(globalAlloc, unicode);
+
+                UnprotectMemory();
+
+                SafeBuffer? bufferToRelease = null;
+                IntPtr ptr = IntPtr.Zero;
+                int byteLength = 0;
+                try
+                {
+                    Span<char> span = AcquireSpan(ref bufferToRelease).Slice(0, _decryptedLength);
+
+                    if (unicode)
+                    {
+                        byteLength = (span.Length + 1) * sizeof(char);
+                    }
+                    else
+                    {
+                        byteLength = Marshal.GetAnsiStringByteCount(span);
+                    }
+
+                    if (globalAlloc)
+                    {
+                        ptr = Marshal.AllocHGlobal(byteLength);
+                    }
+                    else
+                    {
+                        ptr = Marshal.AllocCoTaskMem(byteLength);
+                    }
+
+                    if (unicode)
+                    {
+                        Span<char> resultSpan = new Span<char>((void*)ptr, byteLength / sizeof(char));
```

Comment:
Nit: Division by `sizeof(char)` would be a bit more efficient if the dividend is of `uint` type.

Reply:
I will keep the bit less efficient code here for more readability. I have looked around for other places where we do `(int)value / sizeof(char)` and fixed up a few that are in performance critical code paths.